pkg/auth/dependency/oauth/handler/handler_token.go

```diff
@@ -0,0 +1,389 @@
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/time"
+	"github.com/skygeario/skygear-server/pkg/core/uuid"
+)
+
+// TODO(oauth): write tests
+
+type IDTokenIssuer interface {
+	IssueIDToken(client config.OAuthClientConfiguration, userID string, nonce string) (token string, err error)
+}
+
+type TokenHandler struct {
+	Context context.Context
+	Clients []config.OAuthClientConfiguration
+	Logger  *logrus.Entry
+
+	Authorizations oauth.AuthorizationStore
+	CodeGrants     oauth.CodeGrantStore
+	OfflineGrants  oauth.OfflineGrantStore
+	AccessGrants   oauth.AccessGrantStore
+	Sessions       session.Provider
+	IDTokenIssuer  IDTokenIssuer
+	GenerateToken  TokenGenerator
+	Time           time.Provider
+}
+
+func (h *TokenHandler) Handle(r protocol.TokenRequest) TokenResult {
+	client, errResp := h.resolveClient(r)
+	if errResp != nil {
+		return tokenResultError{Response: errResp}
+	}
+
+	result, err := h.doHandle(client, r)
+	if err != nil {
+		var oauthError *protocol.OAuthProtocolError
+		resultErr := tokenResultError{}
+		if errors.As(err, &oauthError) {
+			resultErr.Response = oauthError.Response
+		} else {
+			h.Logger.WithError(err).Error("authz handler failed")
+			resultErr.Response = protocol.NewErrorResponse("server_error", "internal server error")
+			resultErr.InternalError = true
+		}
+		result = resultErr
+	}
+
+	return result
+}
+
+func (h *TokenHandler) doHandle(
+	client config.OAuthClientConfiguration,
+	r protocol.TokenRequest,
+) (TokenResult, error) {
+	if err := h.validateRequest(r); err != nil {
+		return nil, err
+	}
+
+	switch r.GrantType() {
+	case "authorization_code":
+		return h.handleAuthorizationCode(client, r)
+	case "refresh_token":
+		return h.handleRefreshToken(client, r)
+	default:
+		panic("oauth: unexpected grant type")
+	}
+}
+
+func (h *TokenHandler) resolveClient(r protocol.TokenRequest) (config.OAuthClientConfiguration, protocol.ErrorResponse) {
+	var client config.OAuthClientConfiguration
+	for _, c := range h.Clients {
+		if c.ClientID() == r.ClientID() {
+			client = c
+			break
+		}
+	}
+	if client == nil {
+		return nil, protocol.NewErrorResponse("invalid_client", "invalid client ID")
+	}
+
+	allowedURIs := client.RedirectURIs()
+	redirectURIString := r.RedirectURI()
+	if len(allowedURIs) == 1 && redirectURIString == "" {
+		// Redirect URI is default to the only allowed URI if possible.
+		redirectURIString = allowedURIs[0]
+	}
+
+	allowed := false
+	for _, u := range allowedURIs {
+		if u == redirectURIString {
+			allowed = true
+			break
+		}
+	}
+	if !allowed {
+		return nil, protocol.NewErrorResponse("invalid_request", "redirect URI is not allowed")
+	}
+
+	return client, nil
+}
+
+func (h *TokenHandler) validateRequest(r protocol.TokenRequest) error {
+	switch r.GrantType() {
+	case "authorization_code":
+		if r.Code() == "" {
+			return protocol.NewError("invalid_request", "code is required")
+		}
+		if r.CodeVerifier() == "" {
```

Comment:
For Generic OIDC RP, code verifier is not required. I think we will do it when we introduce client secret later?

https://github.com/SkygearIO/features/blob/fefdd9934bb23d23e13e9b9dc2f340895008b9c7/features/370-auth-ui/oidc.md#generic-oidc-rp

Reply:
Yes, either PKCE or client secret is required.